src/theory/quantifiers/sygus/cegis_unif.cpp

```diff
@@ -235,6 +235,133 @@ void CegisUnif::registerRefinementLemma(const std::vector<Node>& vars,
       NodeManager::currentNM()->mkNode(OR, d_parent->getGuard().negate(), lem));
 }
 
+CegisUnifEnumManager::CegisUnifEnumManager(QuantifiersEngine* qe,
+                                           CegConjecture* parent)
+    : d_qe(qe), d_parent(parent), d_curr_guq_val(qe->getSatContext(), 0)
+{
+  d_tds = d_qe->getTermDatabaseSygus();
+}
+
+void CegisUnifEnumManager::initialize(std::vector<Node>& cs)
+{
+  for (const Node& c : cs)
+  {
+    TypeNode tn = c.getType();
+    d_ce_info[tn].d_candidates.push_back(c);
+  }
+}
+
+void CegisUnifEnumManager::registerEvalPts(std::vector<Node>& eis, Node c)
+{
+  // candidates of the same type are managed
+  TypeNode ct = c.getType();
+  std::map<TypeNode, TypeInfo>::iterator it = d_ce_info.find(ct);
+  Assert(it != d_ce_info.end());
+  it->second.d_eval_points.insert(
+      it->second.d_eval_points.end(), eis.begin(), eis.end());
+  // register at all already allocated sizes
+  for (const std::pair<const unsigned, Node>& p : d_guq_lit)
+  {
+    for (const Node& ei : eis)
+    {
+      Assert(ei.getType() == ct);
+      registerEvalPtAtValue(ct, ei, p.second, p.first);
+    }
+  }
+}
+
+Node CegisUnifEnumManager::getNextDecisionRequest(unsigned& priority)
+{
+  Node lit = getOrMkCurrentLiteral();
+  bool value;
+  if (!d_qe->getValuation().hasSatValue(lit, value))
+  {
+    priority = 0;
+    return lit;
+  }
+  else if (!value)
+  {
+    // propagated false, increment
+    incrementNumEnumerators();
+    return getNextDecisionRequest(priority);
+  }
+  return Node::null();
+}
+
+void CegisUnifEnumManager::incrementNumEnumerators()
+{
+  unsigned new_size = d_curr_guq_val.get() + 1;
+  d_curr_guq_val.set(new_size);
+  Node lit = getOrMkCurrentLiteral();
+  for (std::pair<const TypeNode, TypeInfo>& ci : d_ce_info)
+  {
+    TypeNode ct = ci.first;
+    for (const Node& ei : ci.second.d_eval_points)
+    {
+      registerEvalPtAtValue(ct, ei, lit, new_size);
+    }
+  }
+}
+
+Node CegisUnifEnumManager::getOrMkCurrentLiteral()
+{
+  return getOrMkLiteral(d_curr_guq_val.get());
+}
+
+Node CegisUnifEnumManager::getOrMkLiteral(unsigned n)
+{
+  std::map<unsigned, Node>::iterator itc = d_guq_lit.find(n);
+  if (itc == d_guq_lit.end())
+  {
+    NodeManager* nm = NodeManager::currentNM();
+    // initialize it
+    Node new_lit = Rewriter::rewrite(nm->mkSkolem("G_cost", nm->booleanType()));
+    new_lit = d_qe->getValuation().ensureLiteral(new_lit);
+    AlwaysAssert(!new_lit.isNull());
+    d_qe->getOutputChannel().requirePhase(new_lit, true);
+    d_guq_lit[n] = new_lit;
+    // allocate an enumerator for each candidate
+    for (std::pair<const TypeNode, TypeInfo>& ci : d_ce_info)
```

Comment:
I think it makes more sense for this cade to be ``incrementNumEnumerators`` than here

Reply:
I agree, I've reorganized things a bit to make this clear.